File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxObservableBoolean.java
package xyz.truenight.databinding.rxjava;

import android.databinding.ObservableBoolean;

import java.util.Set;
import java.util.concurrent.CopyOnWriteArraySet;

import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.BehaviorSubject;
import xyz.truenight.utils.Optional;
import xyz.truenight.utils.Utils;

/**
 * Copyright (C) 2017 Mikhail Frolov
 */

public class RxObservableBoolean extends ObservableBoolean {

    private Set<OnPropertyChangedCallback> count = new CopyOnWriteArraySet<>();

    private Disposable subscription;

    private Observable<Optional<Boolean>> observable;

    public RxObservableBoolean(Observable<Optional<Boolean>> observable) {
        super();
        this.observable = observable;
    }

    public RxObservableBoolean(boolean value, Observable<Optional<Boolean>> observable) {
        super(value);
        this.observable = observable;
    }

    public RxObservableBoolean(BehaviorSubject<Optional<Boolean>> subject) {
        super(safe(subject.getValue()));
        this.observable = subject;
    }

    private static Boolean safe(Optional<Boolean> value) {
        return Utils.safe(value == null ? null : value.orElse(null));
    }

    @Deprecated
    @Override
    public void set(boolean value) {
        super.set(value);
    }

    @Override
    public void addOnPropertyChangedCallback(OnPropertyChangedCallback callback) {
        super.addOnPropertyChangedCallback(callback);
        if (count.isEmpty()) {
            subscription = observable.subscribe(value -> super.set(safe(value)));
        }
        count.add(callback);
    }

    @Override
    public void removeOnPropertyChangedCallback(OnPropertyChangedCallback callback) {
        super.removeOnPropertyChangedCallback(callback);
        count.remove(callback);
        if (count.size() == 0) {
            if (subscription != null && !subscription.isDisposed()) {
                subscription.dispose();
            }
        }
    }
}

File: rxjava/src/main/java/xyz/truenight/databinding/rxjava/RxObservableBoolean.kt
package xyz.truenight.databinding.rxjava

import androidx.databinding.ObservableBoolean
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import xyz.truenight.utils.optional.Optional
import xyz.truenight.utils.optional.safe
import java.util.concurrent.CopyOnWriteArraySet

/**
 * Copyright (C) 2017 Mikhail Frolov
 */

class RxObservableBoolean @JvmOverloads internal constructor(val observable: Observable<Boolean>, default: Boolean = false, private val setter: ((Boolean) -> Unit)? = null) : ObservableBoolean(default) {

    private val count = CopyOnWriteArraySet<androidx.databinding.Observable.OnPropertyChangedCallback>()

    private var subscription: Disposable? = null

    override fun set(value: Boolean) {
        super.set(value)
        setter?.let { it(value) }
    }

    override fun addOnPropertyChangedCallback(callback: androidx.databinding.Observable.OnPropertyChangedCallback) {
        super.addOnPropertyChangedCallback(callback)
        if (count.isEmpty()) {
            subscription = observable.subscribe { super.set(it) }
        }
        count.add(callback)
    }

    override fun removeOnPropertyChangedCallback(callback: androidx.databinding.Observable.OnPropertyChangedCallback) {
        super.removeOnPropertyChangedCallback(callback)
        count.remove(callback)
        if (count.isEmpty()) {
            subscription?.dispose()
        }
    }
}

private fun Optional<Boolean>?.safe() = this?.value.safe()

fun Observable<Boolean>.toBinding(default: Boolean = false, setter: ((Boolean) -> Unit)? = null) =
        RxObservableBoolean(this, default, setter)

fun Observable<Optional<Boolean>>.toBinding(default: Optional<Boolean> = Optional.empty(), setter: ((Boolean) -> Unit)? = null) =
        RxObservableBoolean(this.map { it.safe() }, default.safe(), setter)

fun Flowable<Boolean>.toBinding(default: Boolean = false, setter: ((Boolean) -> Unit)? = null) =
        RxObservableBoolean(this.toObservable(), default, setter)

fun Flowable<Optional<Boolean>>.toBinding(default: Optional<Boolean> = Optional.empty(), setter: ((Boolean) -> Unit)? = null) =
        RxObservableBoolean(this.map { it.safe() }.toObservable(), default.safe(), setter)
